src/typesystem.cpp

```diff
@@ -66,30 +114,20 @@ PolyId TypeSystemCore::new_poly(MonoId mono, std::vector<MonoId> vars) {
 	return poly;
 }
 
-TypeFunctionId TypeSystemCore::new_builtin_type_function(int arguments) {
-	TypeFunctionId type_function_var = type_function_header.size();
-
-	type_function_header.push_back({TypeFunctionTag::Builtin, type_function_data.size()});
-	type_function_data.push_back({arguments});
 
-	return type_function_var;
+TypeFunctionId TypeSystemCore::new_builtin_type_function(int arguments) {
+	TypeFunctionId id = m_tf_core.new_term(m_type_functions.size(), {});
+	m_type_functions.push_back({TypeFunctionTag::Builtin, arguments});
+	return id;
 }
 
 TypeFunctionId TypeSystemCore::new_dummy_type_function
     (TypeFunctionTag type, std::unordered_map<std::string, MonoId> structure) {
-	TypeFunctionId type_function_var = type_function_header.size();
-
-	type_function_header.push_back({type, type_function_data.size()});
-	type_function_data.push_back({-1, structure, true});
-
-	return type_function_var;
+	TypeFunctionId id = m_tf_core.new_term(m_type_functions.size(), {});
+	m_type_functions.push_back({type, -1, structure, true});
```

Comment:
Somewhat unrelated, but maybe 0 is better than -1 here? (I mean the argument count)

Reply:
Probably. 0 is more "safe" I think